repr and _*state_dict for qRNN

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31540 repr and _*state_dict for qRNN**

Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
repr and _*state_dict for qRNN

Fixes #31468

Commit:
Adding another method on "repr and _*state_dict for qRNN"


Fixes #31468

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 07c996d4:


* **1/1** broken upstream at merge base 11854bcd (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=11854bcd384be2f2824e075ef7694a95b4709e55&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=07c996d42f2a93e9e8d79778707b9c42a63b36bd).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78529647)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -27,6 +29,30 @@ def __setstate__(self, state):
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                              missing_keys, unexpected_keys, error_msgs):
+        weight = state_dict[prefix + 'weight']
+        self.param = torch.ops.quantized.linear_prepack(weight)
```
I remember `linear_prepack` takes both `weight` and `bias` as argument, did you test this change?

z-a-f replied:
You are right!

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -112,6 +116,34 @@ def extra_repr(self):
         )
 
+    def __repr__(self):
+        # We don't want to show `LinearPackedParams` children, hence custom
```
Can you add a comment here showing how the repr would look for a small example?

raghuramank100 commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -150,6 +180,34 @@ def extra_repr(self):
         return s.format(**self.__dict__)
 
+    def __repr__(self):
+        # We don't want to show `ModuleList` children, hence custom
```
It is hard to understand how the output would look here. Can you add a comment for a simple example, showing what the repr would look like?

raghuramank100 commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -27,6 +29,34 @@ def __setstate__(self, state):
+        return repr(self.unpack())
+
+    def unpack(self):
+        return torch.ops.quantized.linear_unpack(self.param)
```
Please add a test for loading and saving state dicts to verify the functionality here

z-a-f replied:
I created a task for this: #32189

raghuramank100 commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -27,6 +29,34 @@ def __setstate__(self, state):
+                              missing_keys, unexpected_keys, error_msgs):
+        weight = state_dict[prefix + 'weight']
+        bias = state_dict[prefix + 'bias']
+        self.param = torch.ops.quantized.linear_prepack(weight)
```
Still looks like prepack is not fixed?

raghuramank100 reviewed (changes requested):
Few minor changes requested

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -142,6 +146,36 @@ def extra_repr(self):
         )
 
+    def __repr__(self):
+        # We don't want to show `LinearPackedParams` children, hence custom
```
Is this the right approach to take? I think we should show users that there is a LinearPackedParams class and it details instead of hiding it. Particularly with torchbind, there should be no reason to hide this module in the repr

z-a-f replied:
I think it is right -- there is no need for the user to know these details. The repr is usually used just to see the architecture. Showing the `LinearPackedParam` is not very useful. At the same time, if we unpack and show it, it will just cluttern the hole console

Commit:
Update on "repr and _*state_dict for qRNN"


Fixes #31468

Differential Revision: [D19205894](https://our.internmc.facebook.com/intern/diff/D19205894)

z-a-f commented:
Minimum repro example:

```python
import torch
import torch.nn as nn

class LSTMModel(nn.Module):
    """Container module with an encoder, a recurrent module, and a decoder."""

    def __init__(self, ntoken, ninp, nhid, nlayers):
        super(LSTMModel, self).__init__()

        self.rnn = nn.LSTM(ninp, nhid, nlayers)
        self.decoder = nn.Linear(nhid, ntoken)

    def forward(self, input, hidden):
        emb = self.drop(self.encoder(input))
        output, hidden = self.rnn(emb, hidden)
        output = self.drop(output)
        decoded = self.decoder(output)
        return decoded, hidden

model = LSTMModel(
    ntoken = 2,
    ninp = 2,
    nhid = 4,
    nlayers = 2,
)

quantized_model = torch.quantization.quantize_dynamic(
    model, {nn.LSTM, nn.Linear}, dtype=torch.qint8
)

print(quantized_model)
print(quantized_model.rnn)
print(quantized_model.rnn.state_dict())
```

Before this PR:

```console
LSTMModel(
  (rnn): DynamicQuantizedLSTM(
    2, 4, num_layers=2
    (_all_weight_values): ModuleList(
      (0): PackedParameter()
      (1): PackedParameter()
      (2): PackedParameter()
      (3): PackedParameter()
    )
  )
  (decoder): DynamicQuantizedLinear(
    in_features=4, out_features=2, qscheme=torch.per_tensor_affine
    (_packed_params): LinearPackedParams()
  )
)
DynamicQuantizedLSTM(
  2, 4, num_layers=2
  (_all_weight_values): ModuleList(
    (0): PackedParameter()
    (1): PackedParameter()
    (2): PackedParameter()
    (3): PackedParameter()
  )
)
OrderedDict()
```

After this PR:

```console
LSTMModel(
  (rnn): DynamicQuantizedLSTM(2, 4, num_layers=2)
  (decoder): DynamicQuantizedLinear(in_features=4, out_features=2, qscheme=torch.per_tensor_affine)
)
DynamicQuantizedLSTM(2, 4, num_layers=2)
OrderedDict([('_all_weight_values.0.weight', tensor([[-0.4232,  0.0308],
        [ 0.4886,  0.2116],
        [ 0.1577, -0.0269],
        [ 0.0423, -0.1077],
        [-0.2539,  0.0346],
        [-0.3578, -0.3347],
        [ 0.1654,  0.1924],
        [-0.4578, -0.1693],
        [-0.2462, -0.4347],
        [ 0.3809,  0.3039],
        [ 0.4693,  0.2501],
        [-0.3693,  0.4232],
        [-0.4232,  0.2808],
        [-0.0616,  0.2808],
        [ 0.2347, -0.1231],
        [ 0.1885,  0.1193]], size=(16, 2), dtype=torch.qint8,
       quantization_scheme=torch.per_tensor_affine, scale=0.0038470791187137365,
       zero_point=0)), ('_all_weight_values.0.bias', tensor([-0.4421,  0.1431, -0.3968, -0.0575,  0.2676,  0.3340, -0.3531,  0.1048,
        -0.3668, -0.2826, -0.1662,  0.1253,  0.0425, -0.2097, -0.4714,  0.3746],
       requires_grad=True)), ('_all_weight_values.1.weight', tensor([[ 0.4001, -0.4155,  0.2308,  0.3155],
        [ 0.4693, -0.2039, -0.0808, -0.2308],
        [-0.1462, -0.0692,  0.0808, -0.0846],
        [ 0.3809, -0.1847,  0.0692,  0.2616],
        [-0.2039,  0.3347, -0.1308, -0.2077],
        [ 0.1346, -0.0192,  0.0769, -0.0962],
        [ 0.3693,  0.1154, -0.3347, -0.0038],
        [ 0.2270,  0.4193, -0.2308, -0.0115],
        [ 0.4501, -0.0731,  0.4463, -0.4039],
        [-0.3308, -0.1731, -0.0692, -0.4886],
        [-0.4616,  0.2424, -0.3308,  0.4655],
        [ 0.4847, -0.3886,  0.2770, -0.0154],
        [ 0.4001, -0.2424,  0.1039,  0.4616],
        [-0.3155, -0.4616,  0.2847, -0.4693],
        [ 0.0154, -0.1885, -0.4847, -0.0462],
        [-0.3078,  0.1385, -0.0808, -0.3308]], size=(16, 4), dtype=torch.qint8,
       quantization_scheme=torch.per_tensor_affine, scale=0.0038470791187137365,
       zero_point=0)), ('_all_weight_values.1.bias', tensor([-0.4863,  0.3510,  0.0219,  0.4679,  0.3113,  0.3510,  0.1583, -0.1071,
        -0.1239,  0.1534,  0.0960,  0.0927,  0.2397,  0.4424,  0.3677,  0.3715],
       requires_grad=True)), ('_all_weight_values.2.weight', tensor([[-0.0269,  0.1231, -0.2731,  0.4578],
        [ 0.1693, -0.0577,  0.3770,  0.4540],
        [ 0.1000, -0.0385, -0.3385,  0.2347],
        [ 0.3385, -0.2462, -0.3270,  0.4770],
        [-0.1885, -0.2385, -0.3347,  0.0269],
        [ 0.2077,  0.1654, -0.2847,  0.3770],
        [ 0.4386,  0.4616, -0.4193,  0.0115],
        [-0.0923, -0.1654, -0.2116, -0.0846],
        [ 0.0500, -0.0077, -0.4270,  0.2270],
        [-0.1693,  0.3193, -0.3578, -0.1308],
        [ 0.2193, -0.2000,  0.3462,  0.3308],
        [-0.1346,  0.1539,  0.4386,  0.0423],
        [ 0.4116,  0.1693,  0.1885,  0.1423],
        [-0.3847,  0.2154,  0.3809, -0.2154],
        [ 0.0808, -0.3501, -0.3539,  0.1654],
        [-0.1231,  0.1577,  0.2116, -0.2116]], size=(16, 4), dtype=torch.qint8,
       quantization_scheme=torch.per_tensor_affine, scale=0.0038470791187137365,
       zero_point=0)), ('_all_weight_values.2.bias', tensor([ 0.1000, -0.4292,  0.2076,  0.0265,  0.1015, -0.3310,  0.2391, -0.0062,
         0.0778, -0.2181, -0.3969,  0.2349,  0.0224, -0.1896, -0.1292, -0.2441],
       requires_grad=True)), ('_all_weight_values.3.weight', tensor([[-0.4153,  0.3683,  0.2978,  0.0235],
        [-0.0744,  0.0431,  0.4036,  0.3369],
        [-0.3918, -0.5015, -0.0705,  0.4780],
        [-0.0509, -0.1763,  0.1097,  0.2429],
        [ 0.2743, -0.2625,  0.3134, -0.3918],
        [-0.4937,  0.4858, -0.3605,  0.1332],
        [-0.2194, -0.0823, -0.0744, -0.3683],
        [-0.3683, -0.0823,  0.4153, -0.2233],
        [ 0.1920,  0.1332, -0.2233, -0.0862],
        [ 0.4937,  0.2351, -0.4662, -0.0549],
        [ 0.0744,  0.2664, -0.0627, -0.4388],
        [-0.3840,  0.4858, -0.3291,  0.4819],
        [-0.0039, -0.2625, -0.2077, -0.0470],
        [-0.3722,  0.0392, -0.3526, -0.4075],
        [ 0.4349, -0.4623, -0.0666,  0.1410],
        [-0.3409, -0.2233,  0.4388, -0.4858]], size=(16, 4), dtype=torch.qint8,
       quantization_scheme=torch.per_tensor_affine, scale=0.00391801493242383,
       zero_point=0)), ('_all_weight_values.3.bias', tensor([ 0.4249,  0.0961, -0.2984, -0.2992,  0.0977,  0.0327, -0.1053, -0.0706,
         0.0714, -0.2308, -0.1827,  0.0134,  0.0832, -0.3802, -0.3045,  0.2282],
       requires_grad=True))])
```
